File: ALIPHATIC'S ASSIGNMENT/EXERCISE-1/DART/HEAP.dart
/* EXERCISE 1 ON HEAP

Implement a heap data structure storing numeric data in your chosen language with the following operations
heapify() - balance a heap 
insert(item) - add a new item to a heap
find(item) - returns the node containing search item.
findParent(item) - returns parent of an item
findChildren(item) - returns children of an item
remove(item) - remove an item
traverse() - print out values from top to down
heapsort() - sort the items of a heap
*/

import 'dart:io';

class Heap {
  List<int> heap = [];
  int size = 0;

  void heapify() {
    for (int i = size ~/ 2; i >= 1; i--) {
      heapifyDown(i);
    }
  }

  void heapifyDown(int index) {
    int left = 2 * index;
    int right = 2 * index + 1;
    int smallest = index;

    if (left <= size && heap[left] < heap[index]) {
      smallest = left;
    }

    if (right <= size && heap[right] < heap[smallest]) {
      smallest = right;
    }

    if (smallest != index) {
      swap(index, smallest);
      heapifyDown(smallest);
    }
  }

  void insert(int item) {
    heap.add(item);
    size++;
    heapify();
  }

  int find(int item) {
    for (int i = 1; i <= size; i++) {
      if (heap[i] == item) {
        return i;
      }
    }
    return -1;
  }

  int findParent(int item) {
    int index = find(item);
    if (index == -1) {
      return -1;
    }
    return heap[index ~/ 2];
  }

  List<int> findChildren(int item) {
    int index = find(item);
    if (index == -1) {
      return [];
    }
    List<int> children = [];
    if (2 * index <= size) {
      children.add(heap[2 * index]);
    }
    if (2 * index + 1 <= size) {
      children.add(heap[2 * index + 1]);
    }
    return children;
  }

  void remove(int item) {
    int index = find(item);
    if (index == -1) {
      return;
    }
    swap(index, size);
    heap.removeLast();
    size--;
    heapify();
  }

  void traverse() {
    for (int i = 1; i <= size; i++) {
      stdout.write(heap[i].toString() + " ");
    }
    stdout.write("\n");
  }

  void heapsort() {
    for (int i = size; i >= 1; i--) {
      swap(1, i);
      size--;
      heapify();
    }
  }

  void swap(int i, int j) {
    int temp = heap[i];
    heap[i] = heap[j];
    heap[j] = temp;
  }
}

void main() {
  Heap heap = Heap();
  heap.insert(5);
  heap.insert(3);
  heap.insert(7);
  heap.insert(1);
  heap.insert(2);
  heap.insert(6);
  heap.insert(4);
  heap.traverse();
  heap.remove(3);
  heap.traverse();
  print(heap.find(6));
  print(heap.findParent(6));
  print(heap.findChildren(6));
  heap.heapsort();
  heap.traverse();
}

File: ALIPHATIC'S ASSIGNMENT/EXERCISE-1/JAVASCRIPT/HEAP.js
/* EXERCISE 1 ON HEAP

Implement a heap data structure storing numeric data in your chosen language with the following operations
heapify() - balance a heap 
insert(item) - add a new item to a heap
find(item) - returns the node containing search item.
findParent(item) - returns parent of an item
findChildren(item) - returns children of an item
remove(item) - remove an item
traverse() - print out values from top to down
heapsort() - sort the items of a heap
*/

// Heap class

class Heap {
    constructor() {
        this.items = [];
        this.size = 0;
    }

    // heapify() - balance a heap
    heapify() {
        let i = Math.floor(this.size / 2 - 1);
        while (i >= 0) {
            this.percolateDown(i);
            i--;
        }
    }

    // insert(item) - add a new item to a heap
    insert(item) {
        this.items[this.size] = item;
        this.size++;
        this.percolateUp(this.size - 1);
    }

    // find(item) - returns the node containing search item.
    find(item) {
        for (let i = 0; i < this.size; i++) {
            if (this.items[i] === item) {
                return i;
            }
        }
        return -1;
    }

    // findParent(item) - returns parent of an item
    findParent(item) {
        let index = this.find(item);
        if (index === -1) {
            return -1;
        }
        return Math.floor((index - 1) / 2);
    }

    // findChildren(item) - returns children of an item
    findChildren(item) {
        let index = this.find(item);
        if (index === -1) {
            return -1;
        }
        let children = [];
        if (2 * index + 1 < this.size) {
            children.push(this.items[2 * index + 1]);
        }
        if (2 * index + 2 < this.size) {
            children.push(this.items[2 * index + 2]);
        }
        return children;
    }

    // remove(item) - remove an item
    remove(item) {
        let index = this.find(item);
        if (index === -1) {
            return -1;
        }
        this.items[index] = this.items[this.size - 1];
        this.size--;
        this.percolateDown(index);
    }

    // traverse() - print out values from top to down
    traverse() {
        for (let i = 0; i < this.size; i++) {
            console.log(this.items[i]);
        }
    }

    // heapsort() - sort the items of a heap
    heapsort() {
        let sorted = [];
        while (this.size > 0) {
            sorted.push(this
                .items[0]);
            this.items[0] = this.items[this.size - 1];
            this.size--;
            this.percolateDown(0);
        }
        return sorted;
    }

    // percolateUp() - move an item up in the heap
    percolateUp(index) {

        while (index > 0) {
            let parent = Math.floor((index - 1) / 2);
            if (this.items[index] <= this.items[parent]) {
                let temp = this.items[index];
                this.items[index] = this.items[parent];
                this.items[parent] = temp;
                index = parent;
            } else {
                break;
            }
        }
    }

    // percolateDown() - move an item down in the heap
    percolateDown(index) {
        let left = 2 * index + 1;
        let right = 2 * index + 2;
        let min = index;
        if (left < this.size && this.items[left] <= this.items[min]) {
            min = left;
        }
        if (right < this.size && this.items[right] <= this.items[min]) {
            min = right;
        }
        if (min !== index) {
            let temp = this.items[index];
            this.items[index] = this.items[min];
            this.items[min] = temp;
            this.percolateDown(min);
        }
    }
}
